src/lib/support/ReferenceCountedHandle.h

```diff
@@ -0,0 +1,47 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+namespace chip {
+
+/** ReferenceCountedHandle acts like a unique_ptr to an object derived from ReferenceCounted. In contrast to unique_ptr,
+ * the handle will always hold a valid target */
+template <typename Target>
+class ReferenceCountedHandle
+{
+public:
+    explicit ReferenceCountedHandle(Target & target) : mTarget(target) { mTarget.Retain(); }
+
+    ~ReferenceCountedHandle() { mTarget.Release(); }
+
+    ReferenceCountedHandle(const ReferenceCountedHandle & that) : mTarget(that.mTarget) { mTarget.Retain(); }
+
+    ReferenceCountedHandle & operator=(const ReferenceCountedHandle & that) = delete;
+    ReferenceCountedHandle(ReferenceCountedHandle && that)                  = delete;
+    ReferenceCountedHandle & operator=(ReferenceCountedHandle && that) = delete;
+
+    Target & Get() const { return mTarget; }
+    Target * operator->() const { return &mTarget; }
+    bool operator==(const ReferenceCountedHandle & that) const { return mTarget == that.mTarget; }
```

Comment:
This is a somewhat odd implementation of `operator==`.  It tests object equality, not object identity, which is not what I would normally expect operator== on a handle like this to do...

Is this handle just meant to be used on the stack to keep things alive?  If so, I would propose getting rid of the Get() and operator-> and comparison operators, and perhaps deleting the copy constructor and just leaving the constructor and destructor.  And renaming to AutoHoldRef, maybe.  Or AutoRefcount.  Or something like that...  In any case, the "Handle" in the name seems superfluous: the important part is not the handle-ness but the automatic reference counting we get.

We might even want to do that if we do plan to use it in other contexts, and add features as we actually need them and can actually make sure they work correctly because we are using them.

Reply:
This is a mistake when I changed `mTarget` from pointer to reference, I'll fix it.

I agree that copy constructor is not necessary, and I'd prefer an explicit copy function instead of the copy constructor.

But `Get()` and `operator->` are essential for use-cases of  holding an object in another structure. Anyway I'll remove them here, and add them back when needed.